### DISCUSSION TITLE
[Feature Request] possible support for Gdx Ai

### DISCUSSION THREAD

**Discussion by ronjunevaldoz** [libktx/ktx#436] [Ideas]
Hi @czyzby are you planning to have a support for the gdx ai library?

**Comment by czyzby** [comment]
Hey @ronjunevaldoz, I personally haven't used Gdx AI, so I don't feel qualified to provide utilities for it. I'm not against adding a module for another library though, since we already did something similar for VisUI and Ashley. If anyone would like to contribute utilities for Gdx AI, I can do thorough code review and offer some tips in terms of the API design.

**Reply by czyzby** [reply]
gdxAI module will now be available starting from KTX 1.11.0-rc6 thanks to @deviodesign. :)

**Comment by deviodesign** [comment]
Hello @czyzby, I was experimenting a bit and found it could be nice to have a dsl for gdxAI's behavior tree. Let me know what you think. I can find some time to work on the module for ktx.

I'm not sure yet which utilities can be made for other parts like pathfinding and steering though. Suggestions are welcome.

### Example implementation

```kotlin
typealias GdxSequence<E> = com.badlogic.gdx.ai.btree.branch.Sequence<E>

inline fun <E> behaviorTree(rootTask: Task<E>? = null,
                            blackboard: E? = null,
                            init: (@GdxAiTaskDsl BehaviorTree<E>).() -> Unit = {}): BehaviorTree<E> {
    contract { callsInPlace(init, InvocationKind.EXACTLY_ONCE) }
    val behaviorTree = BehaviorTree(rootTask, blackboard)
    behaviorTree.init()
    return behaviorTree
}

inline fun <E> Task<E>.sequence(init: (@GdxAiTaskDsl Sequence<E>).() -> Unit = {}): GdxSequence<E> {
    contract { callsInPlace(init, InvocationKind.EXACTLY_ONCE) }
    val sequence = GdxSequence<E>()
    addChild(sequence)
    sequence.init()
    return sequence
}

// Other Tasks and decorators...

```

### Example usage
```kotlin
class Cat
class Sleep : LeafTask<Cat>()
// etc...

val catBtree = behaviorTree(rootTask = null, blackboard = Cat()) {
    sequence {
        untilSuccess {
            sequence {
                repeat(UniformIntegerDistribution(1, 3), Hunt())
                addChild(Play())
            }
        }
        addChild(Eat())
        addChild(Sleep())
        wait(2f)
        addChild(Drink())
    }
}
```
_It's also possible for users to write their own extension for whatever custom Task, so you don't need addChild(task)._

**Reply by czyzby** [reply]
Unary `+` also seems pretty common as an alternative to `add` in DSLs. This looks good - I haven't used GdxAI personally, but I could try to review the pull request if you'd like to submit your DSL implementation.

**Reply by deviodesign** [reply]
You're right, using unaryPlus seems like a good alternative. It's only possible by subclassing the task implementations though afaik. I'll keep you updated.

**Reply by czyzby** [reply]
I think you might be right, seems like you'd have to extend `Sequence` etc. to enable custom extensions for nested objects. If you don't think it's worth the overhead, feel free to leave the `addChild` syntax as is.

**Reply by deviodesign** [reply]
Alright @czyzby, I created an extension function called `add` instead of unaryPlus. It's short and still allows the use of an init block.

See draft pull request #457.
